numaflow/src/shared/grpc_server.rs

```diff
@@ -0,0 +1,361 @@
+            self.config.server_info_file(),
+            info,
+        )?;
+
+        let (internal_shutdown_tx, internal_shutdown_rx) = mpsc::channel(1);
+        let cln_token = CancellationToken::new();
+
+        // Build the service using the provided builder function
+        let router = service_builder(internal_shutdown_tx, cln_token.clone());
+
+        let shutdown = shutdown_signal(internal_shutdown_rx, shutdown_rx, cln_token);
+
+        router
+            .serve_with_incoming_shutdown(listener, shutdown)
+            .await?;
+
+        Ok(())
+    }
+}
+
+/// Helper function to create a standard server configuration
+pub fn create_server_config(
+    container_type: ContainerType,
+    default_sock_addr: &str,
+    default_server_info_file: &str,
+) -> ServerStarter {
+    ServerStarter::new(container_type, default_sock_addr, default_server_info_file)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tempfile::TempDir;
+
+    #[test]
+    fn test_server_starter_creation() {
+        let starter = ServerStarter::new(ContainerType::Map, "/tmp/test.sock", "/tmp/test-info");
+
+        assert_eq!(
+            starter.socket_file(),
+            std::path::Path::new("/tmp/test.sock")
+        );
+        assert_eq!(
+            starter.server_info_file(),
+            std::path::Path::new("/tmp/test-info")
+        );
+        assert_eq!(starter.max_message_size(), 64 * 1024 * 1024); // 64MB default
+    }
+
+    #[test]
+    fn test_server_starter_configuration() {
+        let tmp_dir = TempDir::new().unwrap();
+        let sock_file = tmp_dir.path().join("custom.sock");
+        let info_file = tmp_dir.path().join("custom-info");
+
+        let starter = ServerStarter::new(ContainerType::Map, "/tmp/test.sock", "/tmp/test-info")
+            .with_socket_file(&sock_file)
+            .with_server_info_file(&info_file)
+            .with_max_message_size(1024)
+            .with_panic_hook(false);
+
+        assert_eq!(starter.socket_file(), sock_file);
+        assert_eq!(starter.server_info_file(), info_file);
+        assert_eq!(starter.max_message_size(), 1024);
+        assert!(!starter.init_panic_hook);
+    }
+
+    #[test]
+    fn test_create_server_config() {
+        let starter = create_server_config(
+            ContainerType::Reduce,
+            "/var/run/numaflow/reduce.sock",
+            "/var/run/numaflow/reducer-server-info",
+        );
+
+        assert_eq!(
+            starter.socket_file(),
+            std::path::Path::new("/var/run/numaflow/reduce.sock")
+        );
+        assert_eq!(
+            starter.server_info_file(),
+            std::path::Path::new("/var/run/numaflow/reducer-server-info")
+        );
+    }
+}
+
+/// Generic gRPC server that can handle any service type
+/// This eliminates the need for duplicate Server implementations across all service files
+#[derive(Debug)]
+pub struct Server<T> {
+    starter: ServerStarter,
+    svc: Option<T>,
```

Comment:
Why is it an option? since `start` and `start_with_shutdown` takes self, can we just borrow it?

Reply:
Also new methods expects svc, means its a mandatory field.

Reply:
we used to do a `take()` earlier, that is why.